### PR TITLE
Fix #22: Hacktoberfest ES6 code refactor

### LINKED ISSUE #22: Refactor code to align to ES6 syntax

### PR DESCRIPTION
**Pull request by Amor64**
This PR contains refactor code to ES6 with following changes : 
1. Conversion of var to const
2. Conversion of promise function to async/await

### CODE REVIEW

File: lib/index.js [hunk 1]
```diff
@@ -3,56 +3,56 @@
...
 
-
   // If the xml content is was not provided by the api client.
   // https://github.com/petkaantonov/bluebird/blob/master/API.md#error-rejectedhandler----promise
   if (!opt.xmlContent) {
-    readFileAsync(opt.filePath, "utf8").then(function(xmlContent) {
-      return xmlContent;
-
-    }).then(_parseWithXml2js).then(function(result) {
+    //Async/wait impl
```
Comment (nsinghal12) [inline lib/index.js:36]: Comment is unnecessary.

File: lib/index.js [hunk 2]
```diff
@@ -97,11 +98,13 @@ function removeSingleArrays(obj) {
...
 
 function readFileAsync(path, encoding) {
-  return new Promise((resolve, reject) => fs.readFile(path, {encoding}, (err, data) => {
-    if (err) {
-      reject(err);
-    } else {
-      resolve(data);
-    }
-  }));
+  return new Promise((resolve, reject) =>
```
Comment (nsinghal12) [inline lib/index.js:101]: This function needs to use async/await as well including at line 102.

File: lib/index.js [hunk 3]
```diff
@@ -61,10 +61,11 @@ module.exports.parse = function(opt, callback) {
  * @param loadedXml {boolean} Whether the xml was loaded from the file-system.
  * @param callback {function} The callback function using Javascript PCS.
  */
+
 function _parseWithXml2js(xmlContent) {
-  return new Promise(function(resolve, reject) {
+  return new Promise(function (resolve, reject) {
```
Comment (nsinghal12) [inline lib/index.js:66]: Request to update to use await instead of a `Promise`.

**Review by nsinghal12 — CHANGES_REQUESTED** [review]
Thanks for the PR. I am requesting minor changes here. Also, please update the title of the PR to include the issue number as a prefix `Fix #22: `.

File: lib/index.js (hunk 3 above)
Comment (Amor64) [reply lib/index.js:66]: @nsinghal12 since async/await is kind of wrapper over Promise. So in order to use async/await on any function, the function should have promise implementation in it. So here **_parseWithXml2js()** implementing promise. That's why we are able to use async/wait in **parse()** in line 22. So here we must use promise so that it can consume as async/await in **parse()**. Let me know your thought on this.

File: lib/index.js (hunk 2 above)
Comment (Amor64) [reply lib/index.js:101]: same as above.

File: lib/index.js (hunk 1 above)
Comment (Amor64) [reply lib/index.js:36]: removed.

File: lib/index.js (hunk 2 above)
Comment (nsinghal12) [reply lib/index.js:97]: See the suggestion on how it's done.

File: lib/index.js [hunk 4]
```diff
@@ -97,13 +94,15 @@ function removeSingleArrays(obj) {
...
-  }));
+  return new Promise((resolve, reject) =>
+    fs.readFile(path, { encoding }, (err, data) => {
+      if (err) {
+        reject(err);
+      } else {
+        resolve(data);
+      }
+    })
+  );
```
Comment (nsinghal12) [inline lib/index.js:105]:
```suggestion
  return await fs.promises.readFile(path, { encoding });
```

File: lib/index.js (hunk 3 above)
Comment (nsinghal12) [reply lib/index.js:62]: Please read my comments and suggestions elsewhere in this review to understand more.

File: test/index.js [hunk 5]
```diff
@@ -1,102 +1,125 @@
...
 
-  describe('when opts is empty', function(){
-    it('parser should throw an error', function(){
-      assert.throws(function(){
-        parse({}, function(err, response){}, Error);
-      });
+  describe("when opts is empty", function () {
+    it("parser should throw an error", async function () {
+      await assertThrowsAsync(
+        async () => await parse({}, function (err, response) {}),
```
Comment (nsinghal12) [inline test/index.js:88]: The `parse` method takes only 2 arguments and not 3.

File: test/index.js [hunk 6]
```diff
@@ -1,102 +1,125 @@
...
 
-  describe('when opts is null', function(){
-    it('parser should throw an error', function() {
-      assert.throws(function(){
-        parse(null, function(err, response){}, Error);
-      });
+  describe("when opts is null", function () {
+    it("parser should throw an error", async function () {
+      await assertThrowsAsync(
+        async () => await parse(null, function (err, response) {}),
```
Comment (nsinghal12) [inline test/index.js:79]: The `parse` method takes only 2 arguments and not 3.

File: test/index.js [hunk 7]
```diff
@@ -1,102 +1,125 @@
-import parse from '../lib/index.js';
-import {expect} from 'chai';
-import assert from 'assert';
-import { dirname } from 'path';
-import { fileURLToPath } from 'url';
+import parse from "../lib/index.js";
```
Comment (nsinghal12) [inline test/index.js:1]: Do not include these changes in the PR. We are not linting/reformatting code in this PR.

File: lib/index.js [hunk 8]
```diff
@@ -2,57 +2,54 @@
...
+  mergeAttrs: true,
 };
 
 /**
  * Parses xml into javascript object by using a file path or an xml content.
  * @param {object} opt Is the option with the filePath or xmlContent and the optional format.
  * @return {object} The pom object along with the timers.
  */
-function parse(opt, callback) {
+const parse = async function (opt, callback) {
```
Comment (nsinghal12) [inline lib/index.js:22]: Use named function syntax here.

**Comment by nsinghal12** [comment]
Hi @Amor64 - I understand how `async`/`await` work. I suggest you re-read on the same and how to convert `Promise`s to make use of them. Any function that returns a `Promise` can be converted to `async` quite easily.

For now I am suggesting the changes for you in PR. Along with them make the below changes:

1. Fix the PR title
2. Add proper description to the PR on what it fixes
3. Modify `readFileAsync` function as:
```js
async function readFileAsync(path, encoding) {
  return await fs.promises.readFile(path, { encoding });
}
```
4.Modify `_parseWithXml2js` function as:
```js
async function _parseWithXml2js(xmlContent) {
  // parse the pom, erasing all
  const pomObject = await xml2js.parseStringPromise(xmlContent, XML2JS_OPTS);

  // Replace the arrays with single elements with strings
  removeSingleArrays(pomObject);

  // Response to the call
  return {
    pomXml: xmlContent, // Only add the pomXml when loaded from the file-system.
    pomObject: pomObject, // Always add the object
  };
}
```
5. Modify the following 2 tests as under:
```js
describe("when opts is null", function () {
    it("parser should throw an error", async function () {
      try {
        await pomParser.parse(null, (err, result) => {});
      } catch(e) {
        expect(e).to.not.be.undefined;
      }
    });
  });
```
and,
```js
describe("when opts is empty", function () {
    it("parser should throw an error", async function () {
      try {
        pomParser.parse({}, (err, result) => {});
      } catch(e) {
        expect(e).to.not.be.undefined;
      }
    });
  });
```
6. Fix up the method JSDocs
7. Format only the lines added/updated. Do not include other linting/reformatting code in this PR.

File: test/xmlcontent.js [hunk 9]
```diff
@@ -1,64 +1,65 @@
-import parse from '../lib/index.js';
-import {expect} from 'chai';
+import parse from "../lib/index.js";
```
Comment (nsinghal12) [inline test/xmlcontent.js:1]: Changes in this file are not relevant to this PR - exclude this file.

File: test/index.js [hunk 10]
```diff
@@ -1,102 +1,125 @@
...
+    before(function (done) {
+      parse({ filePath: POM_PATH }, function (err, response) {
         expect(err).to.be.null;
         expect(response).to.be.an("object");
 
-	pomResponse = response;
-	pom = pomResponse.pomObject;
-	xml = pomResponse.pomXml;
-	done();
+        pomResponse = response;
```
Comment (nsinghal12) [inline test/index.js:39]: Exclude these pure formatting changes from this PR. We will do it as a separate PR.

**Review by nsinghal12 — CHANGES_REQUESTED** [review]

File: lib/index.js (hunk 8 above)
Comment (Amor64) [reply lib/index.js]: done

File: test/xmlcontent.js (hunk 9 above)
Comment (Amor64) [reply test/xmlcontent.js]: this file contains change for conversion of var to let

**Comment by nsinghal12** [comment]
Closing this PR as this has been replaced by #37
